File: 033-ddd/src/main/java/com/demien/ddd/base/Repository.java
package com.demien.ddd.base;

import java.util.List;
import java.util.function.Predicate;

public interface Repository<T> {
    T save(T entity);

    T findById(Long id);

    void delete(Long id);

    List<T> filter(Predicate<T> predicate);
}

File: 047-money-transfer/src/main/scala/com/demien/mtransfer/repo/Repository.scala
package com.demien.mtransfer.repo

trait Repository[T] {

  def getById(id: Int): T

  def save(entity: T): Int

  def update(id: Int, mutator: T => T): Unit

  def biUpdate(id1: Int, mutator1: T => T, id2: Int, mutator2: T => T): Unit

  def find(predicate: T => Boolean): Seq[(Int, T)]

  def history(id: Int): Seq[T]

}
